packages/valory/skills/decision_maker_abci/behaviours/sampling.py

```diff
@@ -103,15 +103,82 @@ def processable_bet(self, bet: Bet, now: int) -> bool:
+        1. Filter out all the bets that have a processed_timestamp != 0 to get a list of new bets.
+        2. If the list of new bets is not empty:
+           2.1 Order the list in decreasing order of liquidity (highest liquidity first).
+           2.2 For bets with the same liquidity, order them in decreasing order of market closing time (openingTimestamp).
+        3. If the list of new bets is empty:
+           3.1 Order the bets in decreasing order of invested_amount.
+           3.2 For bets with the same invested_amount, order them in increasing order of processed_timestamp (least recently processed first).
+           3.3 For bets with the same invested_amount and processed_timestamp, order them in decreasing order of liquidity.
+           3.4 For bets with the same invested_amount, processed_timestamp, and liquidity, order them in decreasing order of market closing time (openingTimestamp).
 
         :param bets: the bets' values to compare for the sampling.
-        :return: the id of the sampled bet, out of all the available bets, not only the given ones.
+        :return: the index of the sampled bet, out of all the available bets, not only the given ones.
         """
-        return self.bets.index(max(bets))
+
+        max_queue_number = max([bet.queue_no for bet in bets])
+
+        if max_queue_number == 0:
+            # Search if any bet is unprocessed
+            new_in_priority_bets = [bet for bet in bets if bet.processed_timestamp == 0]
+
+            if new_in_priority_bets:
+                # Order the list in Decreasing order of liquidity
+                new_in_priority_bets.sort(
+                    key=lambda bet: (bet.scaledLiquidityMeasure, bet.openingTimestamp),
+                    reverse=True,
+                )
+                return self.bets.index(new_in_priority_bets[0])
+            else:
+                # All bets have been processed once, bets can be sampled based on the priority logic
+                bets.sort(
+                    key=lambda bet: (
+                        bet.invested_amount,
+                        -bet.processed_timestamp,  # Increasing order of processed_timestamp
+                        bet.scaledLiquidityMeasure,
+                        bet.openingTimestamp,
+                    ),
+                    reverse=True,
+                )
+                return self.bets.index(bets[0])
+        else:
+            # Check if all bets have processed_timestamp == 0
+            all_bets_not_processed = all(
+                bet.processed_timestamp == 0 for bet in bets if bet.queue_no != -1
+            )
+
+            if all_bets_not_processed:
+                # if none of the bets have been processed, then we should set the current_queue_number to 0
+                # for all none blacklisted bets
+                for bet in self.bets:
+                    if bet.queue_no != -1:
+                        bet.queue_no = 0
+
+                self.store_bets()
+
```

Comment:
Why are you storing here?

Reply:
ahh sorry, I didn't see that async act already has this, will remove this